pact/kpenny/kpenny.pact

```diff
@@ -0,0 +1,241 @@
+
+  (defschema conversion
+    account:string
+    amount-kda:decimal
+    amount-kpenny:decimal)
+
+  (deftable conversions:{conversion})
+
+  (defschema entry
+    balance:decimal
+    guard:guard)
+
+  (deftable ledger:{entry})
+
+  (defcap GOVERNANCE ()
+    (enforce-guard
+      (keyset-ref-guard 'swap-ns-admin)))
+
+  (defcap DEBIT (sender:string)
+    (enforce-guard (at 'guard (read ledger sender))))
+
+  (defcap CREDIT (receiver:string) true)
+
+  (defcap FUND () true)
+
+  (defcap REDEEM () true)
+
+  (defcap TRANSFER:bool
+    ( sender:string
+      receiver:string
+      amount:decimal
+    )
+    @managed amount TRANSFER-mgr
+    (enforce-valid-transfer sender receiver (precision) amount)
+    (compose-capability (DEBIT sender))
+    (compose-capability (CREDIT receiver))
+  )
+
+  (defun TRANSFER-mgr:decimal
+    ( managed:decimal
+      requested:decimal
+    )
+
+    (let ((newbal (- managed requested)))
+      (enforce (>= newbal 0.0)
+        (format "TRANSFER exceeded for balance {}" [managed]))
+      newbal)
+  )
+
+  (defconst MINIMUM_PRECISION:integer 14)
+
+  (defconst CONVERSION_RATE:decimal 1000000.0)
+
+  (defconst KPENNY_BANK:string 'kpenny-bank)
+
+  (defconst SWAP_DEADLINE (time "2021-03-15T00:00:00Z"))
+
+  (defconst FINAL_DEADLINE (time "2021-03-30T00:00:00Z"))
+
+  (defun kpenny-bank-guard () (create-module-guard 'kpenny-admin))
+
+  (defun init ()
+    (coin.create-account KPENNY_BANK (kpenny-bank-guard))
+  )
+
+  (defun enforce-unit:bool (amount:decimal)
+    (enforce-precision (precision) amount))
+
+  (defun create-account:string
+    ( account:string
+      guard:guard
+    )
+    (enforce-valid-account account)
+    (insert ledger account
+      { "balance" : 0.0
+      , "guard"   : guard
+      })
+    )
+
+  (defun get-balance:decimal (account:string)
+    (at 'balance (read ledger account))
+  )
+
+  (defun details:object{fungible-v2.account-details}
+    ( account:string )
+    (with-read ledger account
+      { "balance" := bal
+      , "guard" := g }
+      { "account" : account
+      , "balance" : bal
+      , "guard": g })
+    )
+
+  (defun rotate:string (account:string new-guard:guard)
+    (with-read ledger account
+      { "guard" := old-guard }
+
+      (enforce-guard old-guard)
+
+      (update ledger account
+        { "guard" : new-guard }))
+    )
+
+
+  (defun fund:string (account:string amount:decimal guard:guard)
+    (with-capability (CREDIT account)
+      (require-capability (FUND))
+        (credit account guard amount))
+  )
+
+  (defun convert:string (account:string amount-kda:decimal)
+    (enforce-guard (before-date SWAP_DEADLINE))
+    (coin.transfer account KPENNY_BANK amount-kda)
+    (let
+      ( (tx-id (hash {"account": account, "amount": amount-kda, "salt": (at "block-time" (chain-data))}))
+        (amount-kpenny (* amount-kda CONVERSION_RATE))
+        (g (at 'guard (coin.details account)))
+      )
+      (insert conversions (format "{}-{}" [account, tx-id])
```

Comment:
I'm wondering if a RESERVE event will maybe be enough provenance. Then you can just hit the block explorer to resolve disputes. In any case we should definitely emit the event `(RESERVE account amount-kda)`